subsys/net/ip/tcp.c

```diff
@@ -1346,8 +1346,10 @@ static void tcp_send_zwp(struct k_work *work)
 	tcp_derive_rto(conn);
 
 	if (conn->send_win == 0) {
+		conn->zwp_retries++;
 		(void)k_work_reschedule_for_queue(
-			&tcp_work_q, &conn->persist_timer, K_MSEC(TCP_RTO_MS));
+			&tcp_work_q, &conn->persist_timer,
+			K_MSEC(TCP_RTO_MS << conn->zwp_retries));
```

Comment:
Should we put some upper limit to the doubling, it might be too much to increase the timeout indefinetly, WDYT?

Reply:
I was thinking of it but dropped the idea given that Linux didn't seem to bother. I think we can add some limit however anyway, we'd just need to come up with some reasonable values, as this doesn't seem to be covered by the spec.

How about locking the retransmission counter at the value of 10? This could give ~100 seconds between probes with an initial RTO of 200ms?

Reply:
I wonder why there is the limit of 10, I am fine if we do not check the counter limit. I would prefer that we limit the delay between probes to some max value, not sure what that could be.

Reply:
The send retry and zwp look very similar in behavior if I read it correctly. Is it possible to rework it a little so you can save a timer and counter?

Reply:
> I wonder why there is the limit of 10, I am fine if we do not check the counter limit. I would prefer that we limit the delay between probes to some max value, not sure what that could be.

We can set the max timeout as well, sure. I guess we'd just need to choose some arbitrary value, how about 120 s?

> The send retry and zwp look very similar in behavior if I read it correctly. Is it possible to rework it a little so you can save a timer and counter?

I'm afraid it wouldn't be a trivial change, we'd need to reconfigure the timer every time it's set based on the current context. Plus I'm not really sure if it's possible with the current implementation, I see the following blockers:
*  `send_timer` is used whenever a packet is going to be sent for a local address (corresponding commit: https://github.com/zephyrproject-rtos/zephyr/commit/798588e7092741996a5e27e4d0f411e60a35a606), it'd interfere with probes
* `send_data_timer` is checked and canceled when an ACK is received (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L2181), and if the ACK carried zero window the timer would be already set for ZWP. I'd rather avoid reworking that mechanism not to introduce some regression.

Reply:
In order to merge both functions you might need to add a ZWP next to the RESEND state.

Reply:
> In order to merge both functions you might need to add a ZWP next to the RESEND state.

Can we keep it in a backlog as a possible future improvement? The problem this PR is trying to solve is rather a simple one, I wouldn't like to overcomplicate it with risky changes (I'll OoO for 2 weeks soon).

Reply:
I've pushed a change that adds 2-minute cap on the retry timer.